infra/bisector.py

```diff
@@ -0,0 +1,170 @@
+    engine: The fuzzing engine to be used
+    sanitizer: The fuzzing sanitizer to be used
+    architecture: The system architecture being fuzzed
+    test_case: The file path of the test case that triggers the error
+    fuzzer_name: The name of the fuzzer to be tested
+
+  Returns:
+    The commit SHA that introduced the error or None
+  """
+  local_store_path = 'tmp'
+  repo_url = infer_main_repo(project_name, local_store_path, commit_old)
+  repo_manager = RepoManager(repo_url, local_store_path)
+  commit_list = repo_manager.get_commit_list(commit_old, commit_new)
+  build_fuzzer_from_commit(project_name, commit_list[0], repo_manager.repo_dir,
+                           engine, sanitizer, architecture, repo_manager)
+  error_code = reproduce_impl(project_name, fuzzer_name, False, [], [],
+                              test_case)
+  # Handle the case where there is only one SHA passed in
+  if len(commit_list) == 1:
+    if not error_code:
+      return None
+    return commit_list[0]
+  index = bisection(project_name, 0,
+                    len(commit_list) - 1, commit_list, repo_manager,
+                    len(commit_list), error_code, engine, sanitizer,
+                    architecture, test_case, fuzzer_name)
+  if index is not None:
+    return commit_list[index]
+  return None
+
+
+def bisection(project_name, commit_new_idx, commit_old_idx, commit_list,
+              repo_manager, last_error, error_code, engine, sanitizer,
+              architecture, test_case, fuzzer_name):
+  """Returns the commit ID where a bug was introduced.
+
+  Args:
+    project_name: The name of the oss fuzz project being tested
+    commit_old_idx: The oldest commit SHA index in the search space
+    commit_new_idx: The newest commit SHA index in the search space
+    commit_list: The list of all commit SHAs
+    repo_manager: The class handling all of the git repo calls
+    last_error: The index where the last error was found
+    error_code: The error code of the newest commit(what we are searching for)
+    engine: The fuzzing engine used to fuzz the project
+    sanitizer: THe sanitizer being used for fuzzing
+    architecture: The system architecture being fuzzed
+    test_case: The file path to the test case that caused the error
+    fuzzer_name: The fuzzer that caused the error
+
+  Returns:
+    The index of the commit SHA where the error was introduced
+  """
+  cur_idx = (commit_new_idx + commit_old_idx) // 2
+  build_fuzzer_from_commit(project_name, commit_list[cur_idx],
+                           repo_manager.repo_dir, engine, sanitizer,
+                           architecture, repo_manager)
+  error_exists = (
+      reproduce_impl(project_name, fuzzer_name, False, [], [],
+                     test_case) == error_code)
+  if commit_new_idx == commit_old_idx:
+    if error_exists:
+      return cur_idx
+    return last_error
+
+  if error_exists:
+    return bisection(project_name, cur_idx + 1, commit_old_idx, commit_list,
```

Comment:
I don't understand this `cur_idx + 1` logic, shouldn't bisection look like this:

```
TARGET = 5

def reproduces_on(n):
  print('Trying %d' % n)
  if n >= TARGET:
    return True
  return False


def bisect(old, new):
  print('%d : %d' % (old, new))

  while new - old != 1:
    cur = (old + new) // 2  
    if (reproduces_on(cur)):
      new = cur
    else:
      old = cur

  return new

print('Result: %d ' % bisect(0, 23))

```

Reply:
I mean the `bisect` function in my example, the `reproduces_on` is just an artificial thing that says yes or no

Reply:
I think your example is iterative bisection and I implemented recursive bisection.  cur_idx + 1 is the case where the error is above the most recently checked value. If it was just cur_idx, then the possibility of checking that index again would arise. Same goes for the below case with cur_idx - 1.

Reply:
`def binarySearch (arr, l, r, x): 
  
    # Check base case 
    if r >= l: 
  
        mid = l + (r - l)/2
  
        # If element is present at the middle itself 
        if arr[mid] == x: 
            return mid 
          
        # If element is smaller than mid, then it  
        # can only be present in left subarray 
        elif arr[mid] > x: 
            return binarySearch(arr, l, mid-1, x) 
  
        # Else the element can only be present  
        # in right subarray 
        else: 
            return binarySearch(arr, mid + 1, r, x) 
  
    else: 
        # Element is not present in the array 
        return -1
  `

Reply:
for binary search you need a sorted sequence (i.e. the operators `>` and `==` must be defined), how would you define those for a crash reproduction?

Reply:
They are sorted by time via commit SHA. If the crash reproduces at the specific commit go later into time. If it does not go earlier.

Reply:
The == is why there is a last error term. So if you discover an error and then get all non errors by going earlier into time, you know the most recent error you had was the commit that introduced it.
[0, 0 ,0 , 0, 1, 1]
         ^
[0,1,1]
    ^(also flagged as latest error)
[0]
None
return latest error

Reply:
when I run your implementation on that example, it goes into infinite recursion:

```python
TARGET = 5

def reproduces_on(n):
  print('Trying %d' % n)
  if n >= TARGET:
    return True
  return False

def _bisection(commit_new_idx, commit_old_idx, commit_list, last_error_idx):
  print('%d : %d' % (commit_old_idx, commit_new_idx))
  cur_idx = (commit_new_idx + commit_old_idx) // 2
  error_exists = reproduces_on(cur_idx)
  if commit_new_idx == commit_old_idx:
    if error_exists:
      return cur_idx
    return last_error_idx

  if error_exists:
    return _bisection(cur_idx + 1, commit_old_idx, commit_list, cur_idx)
  if cur_idx == 0:
    return None
  return _bisection(commit_new_idx, cur_idx - 1, commit_list, last_error_idx)


print(_bisection(23, 0, [i for i in range(24)], 24))
```